### PR TITLE
Use django enums

### PR DESCRIPTION
**Pull request by matllubos**

### CODE REVIEW

File: enumfields/enums.py [hunk 1]
```diff
@@ -1,53 +1,86 @@
...
-        obj = BaseEnumMeta.__new__(mcs, name, bases, attrs)
-        for name, choice in choices.items():
-            m = obj[name]
-            for k, v in choice.__dict__.items():
-                if k != 'value':
-                    setattr(m, k, v)
+class ChoiceEnumMeta(ChoicesMeta):
 
-        return unique(obj)
+    def __new__(metacls, classname, bases, classdict):
```
Comment (matllubos) [inline enumfields/enums.py:21]: inspired by code of django enums

File: enumfields/enums.py [hunk 2]
```diff
@@ -1,53 +1,86 @@
...
+def class_to_str(cls):
+    if cls.__class__ == type:
+        return cls
+    else:
+        return '{}.{}'.format(cls.__module__, cls.__name__)
+
+
+class ChoicesEnum(enum.Enum, metaclass=ChoiceEnumMeta):
+
+    def deconstruct_choice(self):
```
Comment (matllubos) [inline enumfields/enums.py:66]: migration serialization to dict

Comment (matllubos) [reply enumfields/enums.py:66]: by default only name, value, next and initial are generated into dict

File: enumfields/enums.py [hunk 3]
```diff
@@ -56,7 +89,11 @@ def __str__(self):
...
+class IntegerChoicesEnum(int, ChoicesEnum):
+    pass
 
-    def __str__(self):  # See Enum.__str__
-        return str(self.label)
+
+class TextChoicesEnum(str, ChoicesEnum):
+
+    def _generate_next_value_(name, start, count, last_values):
+        return name
```
Comment (matllubos) [inline enumfields/enums.py:99]: again django enums inspiration

File: enumfields/fields.py [hunk 4]
```diff
@@ -72,21 +73,45 @@ def _validate_inital_value(self, value, model_instance):
             )
 
 
+def deconstruct_enum(enum):
```
Comment (matllubos) [inline enumfields/fields.py:76]: again helpers for migration serialization

File: enumfields/fields.py [hunk 5]
```diff
@@ -72,21 +73,45 @@ def _validate_inital_value(self, value, model_instance):
...
+
+    return {
+        'name': name,
+        'base': enum_base,
+        'type': enum_type,
+        'choices': choices
+    }
+
+
+def construct_enum(enum):
```
Comment (matllubos) [inline enumfields/fields.py:89]: and deserialization

File: enumfields/fields.py [hunk 6]
```diff
@@ -111,16 +136,20 @@ def refresh_from_db(instance, using=None, fields=None, *args, **kwargs):
         post_save.connect(update_initial_field, sender=cls, weak=False)
 
     def to_python(self, value):
-        if value is None or value == '':
-            return None
         if isinstance(value, self.enum):
             return value
-        for m in self.enum:
```
Comment (matllubos) [inline enumfields/fields.py:118]: this was so so bad. Super dangerous to error

File: enumfields/fields.py [hunk 7]
```diff
@@ -111,16 +136,20 @@ def refresh_from_db(instance, using=None, fields=None, *args, **kwargs):
...
-            if value == m:
-                return m
-            if value == m.value or str(value) == str(m.value) or str(value) == str(m):
```
Comment (matllubos) [inline enumfields/fields.py:121]: str(value) == str(m) -  str of enum value is label. Huh so label is automatically transformed into right enum value. Ouch this is big bug

File: enumfields/fields.py [hunk 8]
```diff
@@ -158,14 +187,14 @@ def get_default(self):
...
-                kwargs['default'] = kwargs['default'].value
+        name, path, args, keywords = super().deconstruct()
+        keywords['enum'] = deconstruct_enum(self.enum)
+        keywords.pop('choices', None)
+        if 'default' in keywords:
+            if hasattr(keywords['default'], 'value'):
+                keywords['default'] = keywords['default'].value
 
-        return name, path, args, kwargs
+        return name, path, args, keywords
```
Comment (matllubos) [inline enumfields/fields.py:197]: in the migrations you can use field.enum to get the last enum in the migration path

File: enumfields/fields.py [hunk 9]
```diff
@@ -192,15 +221,16 @@ def validate(self, value, model_instance):
         super().validate(value, model_instance)
 
 
-class EnumField(EnumFieldMixin, models.CharField):
+class CharEnumField(EnumFieldMixin, models.CharField):
 
     def __init__(self, enum, **kwargs):
-        kwargs.setdefault('max_length', 10)
```
Comment (matllubos) [inline enumfields/fields.py:198]: and what if enum value is longer?

File: enumfields/fields.py [hunk 10]
```diff
@@ -192,15 +221,16 @@ def validate(self, value, model_instance):
         super().validate(value, model_instance)
 
 
-class EnumField(EnumFieldMixin, models.CharField):
```
Comment (matllubos) [inline enumfields/fields.py:195]: this was again dengerous to error. Because EnumField is not parent of NumEnumField

File: enumfields/fields.py [hunk 11]
```diff
@@ -192,15 +221,16 @@ def validate(self, value, model_instance):
...
     def __init__(self, enum, **kwargs):
-        kwargs.setdefault('max_length', 10)
+        enum = construct_enum(enum)
+        kwargs.setdefault('max_length', max(max([len(str(choice.value)) for choice in enum]), 10))
         super().__init__(enum, **kwargs)
         self.validators = []
 
 
-class NumEnumField(EnumFieldMixin, models.IntegerField):
+class IntegerEnumField(EnumFieldMixin, models.IntegerField):
```
Comment (matllubos) [inline enumfields/fields.py:233]: Parent is Integer not Num field..

File: tests/models.py [hunk 12]
```diff
@@ -1,26 +1,25 @@
...
 
-from enumfields import EnumField, NumEnumField, NumEnumSubField
+from enumfields import CharEnumField, IntegerEnumField, IntegerEnumSubField
 
 from .enums import Color, IntegerEnum, LabeledEnum, StateFlow, StateFlowAnyFirst, SubIntegerEnum, Taste, ZeroEnum
 
 
 class MyModel(models.Model):
-    color = EnumField(Color, max_length=1)
-    taste = EnumField(Taste, default=Taste.SWEET)
```
Comment (matllubos) [inline tests/models.py:10]: huh add int enum to string enum field Yuck

File: tests/test_django_models.py [hunk 13]
```diff
@@ -36,29 +36,6 @@ def test_db_value():
     assert cursor.fetchone()[0] == Color.RED.value
 
 
-@pytest.mark.django_db
-def test_enum_int_field_validators():
```
Comment (matllubos) [inline tests/test_django_models.py:40]: not necessary this should be default behaviour

**Review by zamazaljiri — APPROVED** [review]

**Review by rubickcz — APPROVED** [review]

**Pull request merged**